plugin/auth/oauth.go

```diff
@@ -0,0 +1,159 @@
+package auth
+
+// FLOW
+// if not authenticated
+// - button in web settings
+// - redirect to oauth2 provider
+// - redirect back to evcc with code
+// - state authenticated
+// if authenticated
+// - refresh token on demand
+// - save token to db regularly
+
+// STATES
+// not authenticated
+// authenticated
+
+import (
+	"context"
+	"net/http"
+	"strings"
+	"sync"
+
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/server/authredirect"
+	"github.com/evcc-io/evcc/server/db/settings"
+	"github.com/evcc-io/evcc/util"
+	"github.com/evcc-io/evcc/util/request"
+	"golang.org/x/oauth2"
+)
+
+type OAuth struct {
+	mu      sync.Mutex
+	cc      oauth2.Config
+	subject string
+	ts      oauth2.TokenSource
+	cv      string
+	log     *util.Logger
+}
+
+var (
+	mu         sync.Mutex
+	identities = make(map[string]*OAuth)
+)
+
+func getInstance(subject string) *OAuth {
+	return identities[subject]
+}
+
+func addInstance(subject string, identity *OAuth) {
+	identities[subject] = identity
+}
+
+func init() {
+	registry.AddCtx("oauth", NewOauthFromConfig)
+}
+
+func NewOauthFromConfig(ctx context.Context, other map[string]any) (Authorizer, error) {
+	mu.Lock()
+	defer mu.Unlock()
+	// parse oauth config from yaml
+	cc := oauth2.Config{}
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	return NewOauth(cc)
+}
+
+func NewOauth(cc oauth2.Config) (*OAuth, error) {
```

Comment:
```suggestion
func NewOauth(ctx context.Context, oc oauth2.Config) (*OAuth, error) {
```

Reply:
Wenn ich das ändere, dann muss ich den Context in volvo-connected erstellen. 

NewOauthFromConfig habe ich tatsächlich gar nicht getestet, stattdessen rufe ich direkt NewOauth aus volvo-connected auf. Dort habe ich aber eh keinen Context, soll ich den dann dort erstellen oder so lassen?

Reply:
Ich würde sagen background, die Session kann ja mehrere Fahrzeuge bedinen. Also neu erzeugen?